### DISCUSSION TITLE
article/exports-esm-and-cjs/

### DISCUSSION THREAD

**Discussion by giscus** [pengzhanbo/pengzhanbo.cn#37] [Blog-Comment]
# article/exports-esm-and-cjs/

在开发一些公共模块作为一个独立仓库时，有时候可能会在一个使用 es 的项目中通过 import 导入， 有可能在一个 cjs 项目中通过 require 导入。 如何实现单个仓库能够同时被 cjs 和 esm 项目导入呢？ 为什么这么做？ 在过去的时间里，JavaScript 并没有一套标准的模块化系统，并且在过去的时间里，逐渐发展出了各种模块化解决方...

https://pengzhanbo.cn/article/exports-esm-and-cjs/

**Comment by vikiboss** [comment]
👍

**Comment by LetTTGACO** [comment]
文章写的不错，但是有个小问题

> 虽然 Nodejs 从 v14.22.0 版本开始试验性的支持 esm ，并且到 v16 版本，正式支持 esm。 但将库升级到仅支持esm 还是一个比较激进的做法，建议从相对安全的 双格式支持 开始迁移，在合适的时机，过渡到仅支持esm。

这条新闻貌似不太准确，根据nodejs官方文档[nodejs官方文档](https://nodejs.org/docs/latest-v16.x/api/esm.html#modules-ecmascript-modules)的介绍，ESM是从v13.2.0, v12.17.0 v13.2.0、v12.17.0就正式支持了，不用添加实验标记也能运行ESM模块。

**Reply by pengzhanbo** [reply]
确实，而且我发现我把 v14.18.0 写成了 v14.22.0 了。严格来说，低于 V16的版本里，v14.18.0 版本是普遍验证的稳定支持 ESM 的版本，在2022年当时的背景下，一般情况下，一般一个库支持 esm时，标记的node版本兼容是 `^14.18.0 || >=16`。 我应该把文章中的 `试验性支持` 改为 `稳定支持`。
